### ISSUE TITLE
Thoughts on the bash completion API?  (a new shell to run bash completion scripts)

### ISSUE THREAD

**Issue by andychu** [scop/bash-completion#269]
Hi there, I've been experimenting with running bash-completion [1] with my new bash-compatible shell OSH.  I have some things working, but I noticed that the bash-completion code seems to "work around" the bash API, and `git-completion.bash` also copies the workarounds.

I want to get your opinion on the API, and also confirm that I'm understanding the issue correctly.  Long term I would like to find a way to reuse the tens of thousands of lines of `bash-completion` logic for my shell, so I don't have to write it all over again!

----

The main workaround I see is the `__reassemble_comp_words_by_ref` function.  Let me know if there are others.

There seem to be two related issues:

1. bash uses `readline` delimiters (`$COMP_WORDBREAKS`) to tokenize the command line, which you get as `${COMP_WORDS[@]}`.  But this doesn't seem to be what bash-completion or git want, because of cases like `--foo=bar`, where bash will only pass you `bar` because of the `=` token.

2. bash tokenizes the command BEFORE evaluation.   That is, if you have `'sq'` or `"$x"` in your command line, the completion function will get those strings literally!  I believe this is a big mistake, because completion functions aren't really smart enough to understand those tokens.  They would have to essentially have an entire bash parser at their disposal, which even bash does not use!

What do you think about that?

-----

Here's some background, although I've tried to summarize things in this message so you don't have to read a lot.  Thanks for maintaining the bash-completion project!  I have been using it for over a decade in Debian/Ubuntu without appreciating how much work went into it! :)

- October: [Running Bash Completion Scripts with OSH](http://www.oilshell.org/blog/2018/10/10.html) - I can run significant parts of bash-completion (`ls`) and `git-completion.bash`, but there's still many bugs.

- December: [Shell Protocol Designs](http://www.oilshell.org/blog/2018/12/05.html) -- Authors of other shells are interested in making a shell-agnostic completion API.  (This is still very early, and we don't yet agree on what it should look like!)

- Yesterday: [I rewrote interactive completion to use the OSH parser](http://www.oilshell.org/blog/2018/12/16.html#toc_0).  This was a big breakthrough which led me to understand the issues above.

Basically I think that the shell should complete its **own** syntax, and command line tools should complete their own flag syntax as well.  It doesn't make sense to ask command line tools to complete bash's syntax!

[1] https://github.com/oilshell/bash-completion
